src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,903 @@
+      (namespace-price (unwrap! 
+        (compute-namespace-price? namespace)
+        (err ERR_NAMESPACE_BLANK))))
+    ;; The namespace must only have valid chars
+    (asserts!
+      (not (has-invalid-chars namespace))
+      (err ERR_NAMESPACE_CHARSET_INVALID))
+    ;; The namespace must not exist yet in the `namespaces` table
+    (asserts!
+      (is-none (map-get? namespaces ((namespace namespace))))
+      (err ERR_NAMESPACE_ALREADY_EXISTS))
+    ;; The amount burnt must be equal to or greater than the cost of the namespace
+    (asserts!
+      (>= (get stx-burned preorder) namespace-price)
+      (err ERR_NAMESPACE_STX_BURNT_INSUFFICIENT))
+    ;; This transaction must arrive within 24 hours of its `NAMESPACE_PREORDER`
+    (asserts!
+      (< block-height (+ (get created-at preorder) NAMESPACE_PREORDER_CLAIMABILITY_TTL))
+      (err ERR_NAMESPACE_PREORDER_CLAIMABILITY_EXPIRED))
+    ;; The preorder record for this namespace will be marked as "claimed"
+    (map-set namespace-preorders
+      ((hashed-salted-namespace hashed-salted-namespace) (buyer tx-sender))
+      ((created-at (get created-at preorder)) (claimed true) (stx-burned (get stx-burned preorder))))
+    ;; The namespace will be set as "revealed" but not "launched", its price function, its renewal rules, its version,
+    ;; and its import principal will be written to the  `namespaces` table.
+    (map-set namespaces
+      ((namespace namespace))
+      ((namespace-import namespace-import)
+       (revealed-at block-height)
+       (launched-at none)
+       (lifetime lifetime)
+       (price-function price-function)))
+    (ok true))))
+
+;; NAME_IMPORT
+;; Once a namespace is revealed, the user has the option to populate it with a set of names. Each imported name is given
+;; both an owner and some off-chain state. This step is optional; Namespace creators are not required to import names.
+(define-public (name-import (namespace (buff 20))
+                            (name (buff 32))
+                            (beneficiary principal)
+                            (zonefile-hash (buff 20)))
+  (let (
+    (namespace-props (unwrap!
+      (map-get? namespaces ((namespace namespace)))
+      (err ERR_NAMESPACE_NOT_FOUND))))
+      ;; The sender principal must match the namespace's import principal
+      (asserts!
+        (is-eq (get namespace-import namespace-props) tx-sender)
+        (err ERR_NAMESPACE_OPERATION_UNAUTHORIZED))
+      ;; The name's namespace must not be launched
+      (asserts!
+        (is-none (get launched-at namespace-props))
+        (err ERR_NAMESPACE_ALREADY_LAUNCHED))
+      ;; Less than 1 year must have passed since the namespace was "revealed"
+      (asserts!
+        (< block-height (+ (get revealed-at namespace-props) NAMESPACE_LAUNCHABILITY_TTL))
+        (err ERR_NAMESPACE_PREORDER_LAUNCHABILITY_EXPIRED))
+      ;; Mint the new name
+      (try! (mint-or-transfer-name? namespace name beneficiary))
```

Comment:
There needs to be a check here to ensure that the `beneficiary` doesn't own a name yet.  Also, the `owner-name` map should be updated with this name.